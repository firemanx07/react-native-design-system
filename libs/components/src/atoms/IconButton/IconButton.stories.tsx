import { ComponentStory, ComponentMeta } from '@storybook/react';
import React from 'react';

import { StorybookScreen, StorybookSection } from '../../storybook';
import {
  HeartOutlineIcon,
  HeartIcon,
  StarOutlineIcon,
  StarIcon,
  DownloadIcon,
} from '../Icons';
import {
  IconButton as IconButtonBase,
  PropsType as IconButtonProps,
} from './IconButton';

const Story: ComponentMeta<typeof IconButtonBase> = {
  component: IconButtonBase,
  title: 'Atoms/Icon Button',
};
export default Story;

type IconComponentType = (props: {
  width?: number;
  height?: number;
  fill?: string;
}) => JSX.Element;

const Template: ComponentStory<typeof IconButtonBase> = (
  args: Omit<IconButtonProps, 'renderIcon'>,
) => {
  const buttonIcons = [
    HeartOutlineIcon,
    HeartIcon,
    StarOutlineIcon,
    StarIcon,
    DownloadIcon,
  ];

  const renderIcon =
    (Icon: IconComponentType) => (size: number, color: string) =>
      <Icon width={size} height={size} fill={color} />;

  return (
    <StorybookScreen>
      {buttonIcons.map((Icon: IconComponentType, i: number) => (
        <StorybookSection key={i}>
          <IconButtonBase renderIcon={renderIcon(Icon)} {...args} />
        </StorybookSection>
      ))}
    </StorybookScreen>
  );
};

const parameters = {
  controls: {
    include: ['disabled', 'size', 'iconSize'],
  },
};

const defaultArgs: Omit<IconButtonProps, 'renderIcon'> = {
  size: 34,
  iconSize: 20,
  disabled: false,
  onPress: () => {},
};

export const IconButton = Template.bind({});
IconButton.parameters = parameters;
IconButton.args = {
  ...defaultArgs,
};
