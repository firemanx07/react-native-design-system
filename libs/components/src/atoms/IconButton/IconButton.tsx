import { ColorType, styled, useNamespacedTheme } from '@proxym/themes';
import { rgba } from 'polished';
import React, { memo, ReactNode } from 'react';
import { StyleProp, ViewStyle } from 'react-native';

import { TestIDType } from '../../types';
import { ButtonShape } from '../BaseButton/BaseButton.types';

export type PropsType = {
  renderIcon: (size: number, color: string) => ReactNode;
  onPress: () => void;
  size?: number;
  shape?: ButtonShape;
  iconSize?: number;
  iconColor?: ColorType;
  disabled?: boolean;
  style?: StyleProp<ViewStyle>;
  uniColor?: boolean;
} & TestIDType;

export const IconButton = ({
  renderIcon,
  disabled,
  onPress,
  style,
  size = 34,
  iconSize: propsIconSize,
  iconColor: propsIconColor = ColorType.primary,
  shape = ButtonShape.Square,
  uniColor = false,
  testID,
}: PropsType) => {
  const { iconSize: themeIconSize, colors } = useNamespacedTheme();
  const iconSize = propsIconSize || themeIconSize.primary;
  const iconColor = disabled ? rgba(colors.dark, 0.3) : propsIconColor;

  return (
    <Container
      onPress={onPress}
      disabled={disabled}
      size={size}
      shape={shape}
      style={style}
      testID={testID}
      {...(uniColor && { color: iconColor })}
    >
      {renderIcon(iconSize, iconColor)}
    </Container>
  );
};

type ContainerType = {
  size: number;
  shape: ButtonShape;
  color?: string;
};

const Container = styled.TouchableOpacity<ContainerType>`
  width: ${({ size }) => size}px;
  height: ${({ size }) => size}px;
  border-radius: ${({ shape, size }) =>
    shape === ButtonShape.Square ? 7 : size / 2}px;
  background-color: ${({ theme, color }) =>
    rgba(color ?? theme.ds.colors.primaryLight, 0.06)};
  justify-content: center;
  align-items: center;
`;

export default memo(IconButton);
