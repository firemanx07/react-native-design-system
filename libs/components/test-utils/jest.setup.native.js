import { Animated } from 'react-native';

require('react-native-reanimated/lib/reanimated2/jestUtils').setUpTests();

console.warn = () => {};

Animated.timing = () => ({
  start: () => jest.fn(),
});
