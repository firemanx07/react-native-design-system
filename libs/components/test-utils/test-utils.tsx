import { theme, ThemeProvider } from '@celebrate-app/themes';
import { Story } from '@storybook/react';
import {
  cleanup,
  render,
  RenderOptions,
  waitFor,
} from '@testing-library/react-native';
import React, { ReactNode } from 'react';

const Providers = ({ children }: { children: ReactNode }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export const customRender = (
  ui: React.ReactElement<any>,
  options?: RenderOptions
) => {
  return render(ui, { wrapper: Providers, ...options });
};

beforeEach(cleanup);
afterEach(cleanup);

export function testStory(displayName: string, TestComponent: Story) {
  test(`${displayName} snapshot test`, async () => {
    const comp = customRender(<TestComponent {...TestComponent.args} />);
    jest.runOnlyPendingTimers();

    await waitFor(() => {
      expect(comp).toBeTruthy();
      expect(comp.toJSON()).toMatchSnapshot();
    });
  });
}

export function testComponentSnapshot(
  TestComponent: React.ReactElement<
    any,
    string | React.JSXElementConstructor<any>
  >
) {
  test('Snapshot test', () => {
    const comp = customRender(TestComponent);
    expect(comp).toBeTruthy();
    expect(comp.toJSON()).toMatchSnapshot();
  });
}

export function testAllStories(stories: any) {
  Object.keys(stories).forEach((story) => {
    if (story !== 'default') {
      testStory(story, stories[story]);
    }
  });
}
