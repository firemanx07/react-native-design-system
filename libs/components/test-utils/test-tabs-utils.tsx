import { waitFor } from '@testing-library/react-native';

import { customRender } from './test-utils';

type renderTabsProps = {
  elem: JSX.Element;
  testIDs: string[];
};
export const renderTabs = ({ elem, testIDs }: renderTabsProps) => {
  const comp = customRender(elem);
  const tabs: ReturnType<typeof comp.getByTestId>[] = [];
  testIDs.forEach(testID => tabs.push(comp.getByTestId(testID)));
  return { comp, tabs };
};
export const testTabs = async (args: ReturnType<typeof renderTabs>) => {
  await waitFor(() => {
    expect(args.comp).toBeTruthy();
    args.tabs.forEach(tab => expect(tab).toBeDefined());
  });
};
