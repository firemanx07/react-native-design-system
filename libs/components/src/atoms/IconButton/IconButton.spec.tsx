import { testAllStories } from '../../../test-utils';
import * as stories from './IconButton.stories';

describe('IconButton tests', () => {
  testAllStories(stories);
});
