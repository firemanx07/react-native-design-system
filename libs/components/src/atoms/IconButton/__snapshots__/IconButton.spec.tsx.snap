// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton tests IconButton snapshot test 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#F2F3F4",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "minHeight": 600,
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "borderColor": "#222222",
          "borderRadius": 4,
          "borderStyle": "dashed",
          "borderWidth": 1,
          "marginBottom": 24,
          "marginLeft": 24,
          "marginRight": 24,
          "marginTop": 24,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 370,
        },
      ]
    }
    width={370}
  >
    <View
      accessibilityState={
        {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderColor": "#DEDFE0",
          "borderRadius": 17,
          "borderWidth": 1,
          "height": 34,
          "justifyContent": "center",
          "opacity": 1,
          "width": 34,
        }
      }
    >
      <test-file-stub
        fill="#222222"
        height={20}
        width={20}
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "borderColor": "#222222",
          "borderRadius": 4,
          "borderStyle": "dashed",
          "borderWidth": 1,
          "marginBottom": 24,
          "marginLeft": 24,
          "marginRight": 24,
          "marginTop": 24,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 370,
        },
      ]
    }
    width={370}
  >
    <View
      accessibilityState={
        {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderColor": "#DEDFE0",
          "borderRadius": 17,
          "borderWidth": 1,
          "height": 34,
          "justifyContent": "center",
          "opacity": 1,
          "width": 34,
        }
      }
    >
      <test-file-stub
        fill="#222222"
        height={20}
        width={20}
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "borderColor": "#222222",
          "borderRadius": 4,
          "borderStyle": "dashed",
          "borderWidth": 1,
          "marginBottom": 24,
          "marginLeft": 24,
          "marginRight": 24,
          "marginTop": 24,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 370,
        },
      ]
    }
    width={370}
  >
    <View
      accessibilityState={
        {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderColor": "#DEDFE0",
          "borderRadius": 17,
          "borderWidth": 1,
          "height": 34,
          "justifyContent": "center",
          "opacity": 1,
          "width": 34,
        }
      }
    >
      <test-file-stub
        fill="#222222"
        height={20}
        width={20}
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "borderColor": "#222222",
          "borderRadius": 4,
          "borderStyle": "dashed",
          "borderWidth": 1,
          "marginBottom": 24,
          "marginLeft": 24,
          "marginRight": 24,
          "marginTop": 24,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 370,
        },
      ]
    }
    width={370}
  >
    <View
      accessibilityState={
        {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderColor": "#DEDFE0",
          "borderRadius": 17,
          "borderWidth": 1,
          "height": 34,
          "justifyContent": "center",
          "opacity": 1,
          "width": 34,
        }
      }
    >
      <test-file-stub
        fill="#222222"
        height={20}
        width={20}
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "borderColor": "#222222",
          "borderRadius": 4,
          "borderStyle": "dashed",
          "borderWidth": 1,
          "marginBottom": 24,
          "marginLeft": 24,
          "marginRight": 24,
          "marginTop": 24,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 370,
        },
      ]
    }
    width={370}
  >
    <View
      accessibilityState={
        {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderColor": "#DEDFE0",
          "borderRadius": 17,
          "borderWidth": 1,
          "height": 34,
          "justifyContent": "center",
          "opacity": 1,
          "width": 34,
        }
      }
    >
      <test-file-stub
        fill="#222222"
        height={20}
        width={20}
      />
    </View>
  </View>
</View>
`;
